usersession/agent/session_agent.go

```diff
@@ -79,6 +85,70 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	rsp.ServeHTTP(w, r)
 }
 
+type idleTracker struct {
+	mu         sync.Mutex
+	active     map[net.Conn]struct{}
+	lastActive time.Time
+}
+
+var sysGetsockoptUcred = syscall.GetsockoptUcred
+
+func getUcred(conn net.Conn) (*syscall.Ucred, error) {
+	if uconn, ok := conn.(*net.UnixConn); ok {
+		f, err := uconn.File()
+		if err != nil {
+			return nil, err
+		}
+		defer f.Close()
+		return sysGetsockoptUcred(int(f.Fd()), syscall.SOL_SOCKET, syscall.SO_PEERCRED)
+	}
+	return nil, fmt.Errorf("expected a net.UnixConn, but got a %T", conn)
+}
```

Comment:
It would probably make sense to factor out the common code into `netutil`.  We can't import `daemon` here because it pulls in a lot of unwanted dependencies (in particular, dependencies with init functions or global initialisers that would change the behaviour of the `snap` command line tool).